---
layout: post
title: "Has the ship sailed for Microbiome research?"
excerpt: "We use NBCI's esearch API to query for the keyword Microbiome and the results are shocking"
categories: articles
tags: [metagenomics, microbiome, esearch, ncbi]
author: wesley_goi
comments: true
share: true
modified: 2016-12-21
image:
  feature: esearchPublications_9_1.png
  thumb:  meta_thumb.jpg
  credit: Wesley GOI
  creditlink: http://etheleon.github.io
---


Going about my doctoral thesis introduction, the thought occurred to me _perhaps_ the very field I'm writing on about has already lived past its Golden Era, __or has it__?

A knee jerk reaction for me was to see if there's any python or R package which allows me to search the abstracts with the keyword microbiome... this turned up [pubmed_parser](https://github.com/titipata/pubmed_parser/).
However, in order to get this running, I will have to first download a few gigabytes of abstracts in XML from the Open Access subset of pubmed abstracts and run my own pyspark... 

__NOPE! Not going there!__

Then it occurred to me perhaps NCBI has something I could use... in the previous post we talked about the `esearch` API. turns out it could be used.

So below's the script which lets me do this

```R
suppressPackageStartupMessages({
    library(tidyverse)
    library(magrittr)
    library(rvest) # for XML
    library(artyfarty) #because theme_bw is too boring
})
```

# NCBI ESEARCH

There's the NBCI's esearch [API](https://www.ncbi.nlm.nih.gov/books/NBK25499/)

Within it, there's the date range options

#### mindate, maxdate API filter

Date range used to limit a search result by the date specified by datetype. These two parameters (mindate, maxdate) must be used together to specify an arbitrary date range. The general date format is YYYY/MM/DD, and these variants are also allowed: YYYY, YYYY/MM.

So we will be searching between _1997_ to _2017_, a 20 year period.

So lets begin .....

#### Keyword: Microbiome

with other synonyms microbiota


```R
#microbiome
searchTerm="https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=%s&mindate=%s&maxdate=%s"

keyword = "microbiome"
df = mapply(function(start, end){
            count = read_xml(sprintf(searchTerm, keyword, start, end)) %>%
            as_list %$%
            Count %>%
            unlist
       tibble(count, start, end)
},  start = 1997:2016,
    end = 1998:2017,
    SIMPLIFY=FALSE
) %>% do.call(rbind,.)

```

#### Keyword: Cancer

Used as a comparison


```R
keyword = "cancer"

df2 = mapply(function(start, end){
            count = read_xml(sprintf(searchTerm, keyword, start, end)) %>%
            as_list %$%
            Count %>%
            unlist
       tibble(count, start, end)
},  start = 1997:2016,
    end = 1998:2017,
    SIMPLIFY=FALSE
) %>% do.call(rbind,.)
```

Putting the two together before we start plotting


```R
df %<>% setNames(c("microbiome", "start", "end"))
df %<>% mutate(cancer = as.integer(df2$count))
```


```R
df$microbiome %<>% as.integer
df %<>% select(start, end, microbiome, cancer)
df
```
As you can the see the order is slighly different between the two,
you'll probably have to do some scaling.

| start | end  | microbiome | cancer |
| ---   | ---  | ---        | ---    |
| 1997  | 1998 | 91         | 116522 |
| 1998  | 1999 | 110        | 124613 |
| 1999  | 2000 | 133        | 131481 |
| 2000  | 2001 | 149        | 139577 |
| 2001  | 2002 | 196        | 153651 |
| 2002  | 2003 | 249        | 166393 |
| 2003  | 2004 | 304        | 170676 |
| 2004  | 2005 | 419        | 181504 |
| 2005  | 2006 | 576        | 190710 |
| 2006  | 2007 | 744        | 198618 |
| 2007  | 2008 | 955        | 210488 |
| 2008  | 2009 | 1285       | 219686 |
| 2009  | 2010 | 1741       | 231079 |
| 2010  | 2011 | 2610       | 248046 |
| 2011  | 2012 | 3899       | 265171 |
| 2012  | 2013 | 5607       | 281240 |
| 2013  | 2014 | 8211       | 308483 |
| 2014  | 2015 | 10951      | 331775 |
| 2015  | 2016 | 13439      | 331631 |
| 2016  | 2017 | 14058      | 285408 |


Since the 2.2.0 version of ggplot2, Hadley has included the `sec_axid` function in the library which lets you add a secondary axis as long as it's amenable to a straight forward transformation.


```R
ggplot(df, aes(x=end)) +
    geom_line(aes(y=microbiome, color="Microbiome"), size=1.1) +
    geom_line(aes(y=cancer/20, color="Cancer"), size=1.5, linetype="dotted") + 
    # manipulated the cancer values by dividing by 20
    scale_x_continuous(breaks=1998:2017)+
    scale_y_continuous(sec.axis = sec_axis(~.*20, name = "Number of Publications [Cancer]"))+ 
    # restores the division
    # lets we set the axis title
    scale_color_manual("Search Terms",values = pal("five38"))+
    theme_scientific() +
    theme(axis.text.x=element_text(angle=90), 
          legend.position = c(0.9, 0.2)) +
    xlab("Year") + ylab("Number of Publications [Microbiome]")
```
![publications-with-keyword-microbiome](/images/esearchPublications_9_1.png)

There you have it guys.
The growth in publications/articles revolving around microbiome or at least associated to it have been growing at breakneck pace faster than cancer, almost exponential.

Hope this will be helpful for all future students who'll be add this kind of plots to their introduction.
