---
layout: post
title: "Has the ship sailed for Microbiome research?"
excerpt: "We use NBCI's esearch API to query for the keyword Microbiome and the results are shocking"
categories: articles
tags: [metagenomics, microbiome, esearch, ncbi]
author: wesley_goi
comments: true
share: true
modified: 2016-12-21
image:
  feature: coverIntro.png
  thumb:  meta_thumb.jpg
  credit: Wesley GOI
  creditlink: http://etheleon.github.io
---


Going about my doctoral thesis introduction chapter, the thought occurred to me _perhaps_ the very field I'm writing about has already lived past its Golden Era, or has it?

Well, a knee jerk reaction was to see if there's any python or R package for me to search the keywords associated with pubmed abstracts... this turned up [pubmed_parser](https://github.com/titipata/pubmed_parser/)
But to get this running, I had to download a few gigabytes of abstracts in XML from the Open Access subset of pubmed and run my own pyspark... NOPE! Not going there.

Then it occurred to me perhaps NCBI has something I could use... in the previous post we talked about the `esearch` API. turns out it could be used.

```R
suppressPackageStartupMessages({
    library(tidyverse)
    library(magrittr)
    library(rvest) # for XML
    library(artyfarty)
})
```

# ESEARCH

There's the NBCI's esearch [API](https://www.ncbi.nlm.nih.gov/books/NBK25499/)

Within it, there's the date range options

## mindate, maxdate

Date range used to limit a search result by the date specified by datetype. These two parameters (mindate, maxdate) must be used together to specify an arbitrary date range. The general date format is YYYY/MM/DD, and these variants are also allowed: YYYY, YYYY/MM.

So we are searching between _1997_ to _2017_

So lets begin .....

## Keyword: Microbiome

with other synonyms microbiota


```R
#microbiome
keyword = "microbiome"
df = mapply(function(start, end){
            count = read_xml(sprintf("https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=%s&mindate=%s&maxdate=%s", keyword, start, end)) %>%
            as_list %$%
            Count %>%
            unlist
       tibble(count, start, end)
},  start = 1997:2016, 
    end = 1998:2017,
    SIMPLIFY=FALSE
) %>% do.call(rbind,.)

```

## Keyword: Cancer

Used as a comparison


```R
keyword = "cancer"

df2 = mapply(function(start, end){
            count = read_xml(sprintf("https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=%s&mindate=%s&maxdate=%s", keyword, start, end)) %>%
            as_list %$%
            Count %>%
            unlist
       tibble(count, start, end)
},  start = 1997:2016, 
    end = 1998:2017,
    SIMPLIFY=FALSE
) %>% do.call(rbind,.)
```

Putting the two together before we start plotting


```R
df %<>% setNames(c("microbiome", "start", "end"))
df %<>% mutate(cancer = as.integer(df2$count))
```


```R
df$microbiome %<>% as.integer
df %<>% select(start, end, microbiome, cancer)
df
```


<table>
<thead><tr><th scope=col>start</th><th scope=col>end</th><th scope=col>microbiome</th><th scope=col>cancer</th></tr></thead>
<tbody>
	<tr><td>1997  </td><td>1998  </td><td>   91 </td><td>116522</td></tr>
	<tr><td>1998  </td><td>1999  </td><td>  110 </td><td>124613</td></tr>
	<tr><td>1999  </td><td>2000  </td><td>  133 </td><td>131481</td></tr>
	<tr><td>2000  </td><td>2001  </td><td>  149 </td><td>139577</td></tr>
	<tr><td>2001  </td><td>2002  </td><td>  196 </td><td>153651</td></tr>
	<tr><td>2002  </td><td>2003  </td><td>  249 </td><td>166393</td></tr>
	<tr><td>2003  </td><td>2004  </td><td>  304 </td><td>170676</td></tr>
	<tr><td>2004  </td><td>2005  </td><td>  419 </td><td>181504</td></tr>
	<tr><td>2005  </td><td>2006  </td><td>  576 </td><td>190710</td></tr>
	<tr><td>2006  </td><td>2007  </td><td>  744 </td><td>198618</td></tr>
	<tr><td>2007  </td><td>2008  </td><td>  955 </td><td>210488</td></tr>
	<tr><td>2008  </td><td>2009  </td><td> 1285 </td><td>219686</td></tr>
	<tr><td>2009  </td><td>2010  </td><td> 1741 </td><td>231079</td></tr>
	<tr><td>2010  </td><td>2011  </td><td> 2610 </td><td>248046</td></tr>
	<tr><td>2011  </td><td>2012  </td><td> 3899 </td><td>265171</td></tr>
	<tr><td>2012  </td><td>2013  </td><td> 5607 </td><td>281240</td></tr>
	<tr><td>2013  </td><td>2014  </td><td> 8211 </td><td>308483</td></tr>
	<tr><td>2014  </td><td>2015  </td><td>10951 </td><td>331775</td></tr>
	<tr><td>2015  </td><td>2016  </td><td>13439 </td><td>331631</td></tr>
	<tr><td>2016  </td><td>2017  </td><td>14058 </td><td>285408</td></tr>
</tbody>
</table>




```R
ggplot(df, aes(x=end)) +
    geom_line(aes(y=microbiome, color="Microbiome"), size=1.1) +
    geom_line(aes(y=cancer/20, color="Cancer"), size=1.5, linetype="dotted") +
    scale_x_continuous(breaks=1998:2017)+
    scale_y_continuous(sec.axis = sec_axis(~.*20, name = "Number of Publications [Cancer]"))+
    scale_color_manual("Search Terms",values = pal("five38"))+
    theme_scientific() +
    theme(axis.text.x=element_text(angle=90), 
          legend.position = c(0.9, 0.2)) +
    xlab("Year") + ylab("Number of Publications [Microbiome]")
```



![png](esearchPublications_9_1.png)


There you have it guys.

The growth in publications/articles revolving around microbiome or at least associated to it have been growing at breakneck pace faster than cancer
